DOCSP-28586: How to use section (#341)
mongodb/docs-java: source/fundamentals/aggregation/aggregation-expression-operations.txt

   import static com.mongodb.client.model.mql.MqlValues.*;
   import static java.util.Arrays.asList;

To access document fields in an aggregation expression, you need to reference the
document being processed by the aggregation pipeline. Use the
``current()`` method to refer to this document.

> To access the value of a field, you must use the appropriately typed

Reviewer: [suggestion]
I'd consider combining this with the paragraph above about using `current()` since they are closely tied together. It would also make the previous paragraph feel less like it is free floating.

Author: agreed-- there are many short paragraphs in this section

method, such as ``getString()`` or ``getDate()``. The following code
shows how to reference a string field called ``name``:

.. code-block:: java
   :copyable: true
   
   current().getString("name")

To specify a value in an operation, pass it to the ``of()`` method to
convert it to a valid type. The following code shows how to reference a
value of ``1.0``:

.. code-block:: java
   :copyable: true
   
   of(1.0)

To create an operation, chain a method to your field or value reference.
You can build more complex operations by chaining additional methods.

The following example creates an operation to find patients in New
Mexico who have visited the doctor’s office at least once. The operation
performs the following actions:

- Checks if the size of the ``visitDates`` array is greater than ``0``
  by using the ``gt()`` method 
- Checks if the ``state`` field value is “New Mexico” by using the
  ``eq()`` method

The ``and()`` method links these operations so that the pipeline stage

> matches only documents that meet both criteria.

Reviewer: [suggestion]
Consider combining this above in the bulleted list. Something like

"Combines these two operations by using the `and()` method". 

If you do that, you might need to slightly reword the introductory sentence though.

Author: I think the way it is now is a bit easier to follow, in the other case the last bullet point would reference the previous two, which I think is more challenging. For me, it's easier to understand this conceptually as two separate operations that are then linked

.. code-block:: java
   :copyable: true
      
   current()
         .getArray("visitDates")
         .size()
         .gt(of(0))
         .and(current()
                 .getString("state")
                 .eq(of("New Mexico")));

In most cases, you need to include your operation in a method such as
``computed()`` or ``expr()`` to process it for use in an aggregation
pipeline stage.

> 

Reviewer: [question]
Do we want to explain at all why you might use computed or expr? Right now it is left as an exercise to the reader to figure out why one might want to use one or the other.

Author: Chris, Maxim, and I met yesterday and spent a lot of time thinking/talking about this. We came to the conclusion that it might be going into the weeds too much to explain that a type conversion takes place, etc. etc. Easier just to say that users may need to use these methods to make the builders methods accept the expressions.

To create a usable aggregation pipeline stage, include your expression
in an aggregates builder method such as ``match(<expression>)`` or
``project(<expression>)``. To learn more about these methods, see
:ref:`Aggregates Builders <aggregates-builders>`.

The examples use the ``Arrays.asList()`` method to create a list of
aggregation stages that is passed to the ``aggregate()`` method of
``MongoCollection``.

Methods such as ``computed()`` and ``expr()`` that take values of type
``TExpression`` allow you to use your expressions in aggregations. For
example, the ``expr()`` method processes your operation as an expression

> that the ``match()`` method will accept.

Author: I wonder if this should be in a note at line 96 instead.

Reviewer: Considering my earlier comment, I would be inclined to agree. Putting this down here creates a disconnect from the knowledge. I don't know if it needs to be a tip, but I would definitely move it up.

Author: made a note, LMK if you think the flow of info is good this way or if the content should be included as regular text.

Reviewer: I think this works just fine as a note.

   import static com.mongodb.client.model.mql.MqlValues.*;
   import static java.util.Arrays.asList;

To access document fields in an expression, you need to reference the
document being processed by the aggregation pipeline. Use the
``current()`` method to refer to this document. To access the value of a
field, you must use the appropriately typed method, such as
``getString()`` or ``getDate()``. The following code shows how to
reference a string field called ``name``:

.. code-block:: java
   :copyable: true
   
   current().getString("name")

To specify a value in an operation, pass it to the ``of()`` method to
convert it to a valid type. The following code shows how to reference a
value of ``1.0``:

.. code-block:: java
   :copyable: true
   

>    of(1.0)

Author: thanks for pointing that out! I think I'll have the chance to show `ofArray()` in the array operations section later on.

To create an operation, chain a method to your field or value reference.
You can build more complex operations by chaining additional methods.

The following example creates an operation to find patients in New
Mexico who have visited the doctor’s office at least once. The operation
performs the following actions:

- Checks if the size of the ``visitDates`` array is greater than ``0``
  by using the ``gt()`` method 
- Checks if the ``state`` field value is “New Mexico” by using the
  ``eq()`` method

The ``and()`` method links these operations so that the pipeline stage
matches only documents that meet both criteria.

.. code-block:: java
   :copyable: true
      
   current()
         .getArray("visitDates")
         .size()
         .gt(of(0))
         .and(current()
                 .getString("state")
                 .eq(of("New Mexico")));

In most cases, you need to include your operation in a method such as
``computed()`` or ``expr()`` to process it for use in an aggregation

> pipeline stage.

Author: ok -- removing this para and taking the text out of the note!

> pipeline stage.

Author: adapted the text for flow

.. note::

   Methods such as ``computed()`` and ``expr()`` that take values of
   type ``TExpression`` allow you to use your expressions in
   aggregations. For example, the ``expr()`` method processes your

>    operation as an expression that the ``match()`` method will accept.

Author: disregard my reply on the previous PR comment -- how is this paragraph read now with revisions?

>    operation as an expression that the ``match()`` method will accept.

Reviewer 2: Looks great!

> To create a usable aggregation pipeline stage, include your expression
> in an aggregates builder method such as ``match(<expression>)`` or
> ``project(<expression>)``. To learn more about these methods, see
> :ref:`Aggregates Builders <aggregates-builders>`.

Reviewer 2: The project and match methods do not take `<expression>`s directly. They accept plain BSON. That BSON can be in a format that is unrelated to expressions (like match's `field: 0 or 1` format). The `$expr` operator indicates that its contents are indeed an expression.

```suggestion
To create a usable aggregation pipeline stage, include your expression
in an aggregates builder method, such as in ``match(expr(<expression>))`` or
``project(fields(computed("<field name>", <expression>)))``. To learn more about these methods, see
:ref:`Aggregates Builders <aggregates-builders>`.
```

   import static com.mongodb.client.model.mql.MqlValues.*;
   import static java.util.Arrays.asList;

To access document fields in an expression, you need to reference the
document being processed by the aggregation pipeline. Use the
``current()`` method to refer to this document. To access the value of a
field, you must use the appropriately typed method, such as
``getString()`` or ``getDate()``. When you specify the type for a field,

> you ensure that the driver recognizes its type correctly and lets you
> use methods that are compatible with that type. The following code

Reviewer 2: ```suggestion
you ensure that the driver provides only those
methods that are compatible with that type. The following code
```
(Avoiding the "recognition" (checking) and "letting" concepts since they are mildly misleading - for example, driver can't check types.)

  by using the ``gt()`` method 
- Checks if the ``state`` field value is “New Mexico” by using the
  ``eq()`` method

The ``and()`` method links these operations so that the pipeline stage
matches only documents that meet both criteria.

.. code-block:: java
   :copyable: true
      
   current()
         .getArray("visitDates")
         .size()
         .gt(of(0))
         .and(current()
                 .getString("state")
                 .eq(of("New Mexico")));

While some aggregation stages, such as ``group()``, accept operations
directly, other stages expect that you first include your operation in a
method such as ``computed()`` or ``expr()``. These methods, which take
values of type ``TExpression``, allow you to use your expressions in
certain aggregations.

To complete your aggregation pipeline stage, include your expression
in an aggregates builder method. The following list provides formats for
how to include your expression in common aggregates builder methods:

- ``match(expr(<expression>))``
- ``project(fields(computed("<field name>", <expression>)))``
- ``group(<expression>)``

To learn more about these methods, see

> :ref:`Aggregates Builders <aggregates-builders>`.

Author: the first sentence in the overview for the Aggregates Builders page (which is linked here) has a link to the API documentation!
